File: tests/fixtures/user/linux/src/kcmp.c
/* kcmp (kernel/kcmp.c) and the objects tasks share by their clone flags:
 * the checks in order (tasks, then the right to inspect them, then the
 * type), open file descriptions (duplicates, O_PATH, the index as an
 * unsigned int, a total order), what threads share (address space,
 * tables, signal handlers) and what they share only by CLONE_IO or
 * CLONE_SYSVSEM (an I/O context, a semaphore undo list, both absent until
 * made), epoll items by descriptor and offset, a leader that exited while
 * its threads run, and init, which another user may not inspect. The
 * threads are made by a bare clone (rawthread.h). Root drops to nobody
 * first. */
#define _GNU_SOURCE
#include <fcntl.h>
#include <pthread.h>
#include <sched.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/epoll.h>
#include <sys/syscall.h>
#include <sys/wait.h>
#include <unistd.h>
#include <linux/futex.h>
#include "check.h"
#include "rawthread.h"

#define KCMP_FILE 0
#define KCMP_VM 1
#define KCMP_FILES 2
#define KCMP_FS 3
#define KCMP_SIGHAND 4
#define KCMP_IO 5
#define KCMP_SYSVSEM 6
#define KCMP_EPOLL_TFD 7
#define BAD ((void *)16)
#define THREAD (CLONE_VM | CLONE_FS | CLONE_FILES | CLONE_SIGHAND | CLONE_THREAD)
#define IOPRIO_BE (2 << 13)

static long kcmp(pid_t a, pid_t b, int type, unsigned long i1, unsigned long i2) {
    return syscall(SYS_kcmp, a, b, type, i1, i2);
}

/* Distinct objects: 1 or 2, the other way round with the arguments. */
static int ordered(pid_t a, pid_t b, int type, unsigned long i1, unsigned long i2) {
    long x = kcmp(a, b, type, i1, i2), y = kcmp(b, a, type, i2, i1);
    return (x == 1 && y == 2) || (x == 2 && y == 1);
}

static void order_and_files(void) {
    pid_t me = getpid();
    CHECK_ERR("no-task-first", kcmp(99999, me, 99, 0, 0), ESRCH);
    CHECK_ERR("no-task-second", kcmp(me, 0, 99, 0, 0), ESRCH);
    CHECK_ERR("unknown-type", kcmp(me, me, 8, 0, 0), EINVAL);
    CHECK_ERR("negative-type", kcmp(me, me, -1, 0, 0), EINVAL);
    int a = open("/", O_RDONLY), b = open("/", O_RDONLY), d = dup(a);
    int p = open("/", O_PATH);
    CHECK("file-dup", kcmp(me, me, KCMP_FILE, a, d) == 0);
    CHECK("file-distinct", ordered(me, me, KCMP_FILE, a, b));
    CHECK("file-o-path", ordered(me, me, KCMP_FILE, a, p));
    CHECK("file-index-unsigned-int", kcmp(me, me, KCMP_FILE, a, (1UL << 32) | d) == 0);
    CHECK_ERR("file-none", kcmp(me, me, KCMP_FILE, a, 999), EBADF);
    CHECK_ERR("file-negative", kcmp(me, me, KCMP_FILE, -1UL, a), EBADF);
    /* Three descriptions in a total order. */
    int c = open("/", O_RDONLY);
    int v[3] = {a, b, c};
    for (int i = 0; i < 3; i++)
        for (int j = 0; j + 1 < 3 - i; j++)
            if (kcmp(me, me, KCMP_FILE, v[j], v[j + 1]) == 2) {
                int t = v[j];
                v[j] = v[j + 1];
                v[j + 1] = t;
            }
    CHECK("file-total-order", kcmp(me, me, KCMP_FILE, v[0], v[1]) == 1 &&
                                  kcmp(me, me, KCMP_FILE, v[1], v[2]) == 1 &&
                                  kcmp(me, me, KCMP_FILE, v[0], v[2]) == 1);
    close(a);
    close(b);
    close(c);
    close(d);
    close(p);
}

/* Bare threads that wait until told to stop. */
static int stop;
static char stacks[4][1 << 14] __attribute__((aligned(16)));
static int ctids[4];

static int waiter(void *arg) {
    (void)arg;
    while (!__atomic_load_n(&stop, __ATOMIC_ACQUIRE))
        syscall(SYS_futex, &stop, FUTEX_WAIT, 0, NULL);
    return 0;
}

static pid_t spawn(int i, unsigned long flags) {
    ctids[i] = 1;
    return raw_thread(flags | CLONE_CHILD_CLEARTID, stacks[i] + sizeof stacks[i], waiter, NULL,
                      &ctids[i]);
}

static void stop_all(void) {
    __atomic_store_n(&stop, 1, __ATOMIC_RELEASE);
    syscall(SYS_futex, &stop, FUTEX_WAKE, 4, NULL);
    for (int i = 0; i < 4; i++)
        while (__atomic_load_n(&ctids[i], __ATOMIC_ACQUIRE))
            syscall(SYS_futex, &ctids[i], FUTEX_WAIT, ctids[i], NULL);
}

static void threads(void) {
    pid_t me = getpid();
    pid_t t = spawn(0, THREAD);
    CHECK("thread", t > 0);
    CHECK("thread-vm", kcmp(me, t, KCMP_VM, 0, 0) == 0);
    CHECK("thread-files", kcmp(me, t, KCMP_FILES, 0, 0) == 0);
    CHECK("thread-fs", kcmp(me, t, KCMP_FS, 0, 0) == 0);
    CHECK("thread-sighand", kcmp(me, t, KCMP_SIGHAND, 0, 0) == 0);
    CHECK("thread-no-io-contexts", kcmp(me, t, KCMP_IO, 0, 0) == 0);
    CHECK("thread-no-undo-lists", kcmp(me, t, KCMP_SYSVSEM, 0, 0) == 0);
    CHECK("ioprio-set", syscall(SYS_ioprio_set, 1, 0, IOPRIO_BE | 4) == 0);
    CHECK("io-context-own", ordered(me, t, KCMP_IO, 0, 0));
    pid_t io = spawn(1, THREAD | CLONE_IO);
    CHECK("io-context-shared", kcmp(me, io, KCMP_IO, 0, 0) == 0);
    pid_t copy = spawn(2, THREAD);
    CHECK("io-context-copied", ordered(me, copy, KCMP_IO, 0, 0));
    CHECK("io-context-copied-priority", syscall(SYS_ioprio_get, 1, copy) == (IOPRIO_BE | 4));
    pid_t sem = spawn(3, THREAD | CLONE_SYSVSEM);
    CHECK("undo-list-shared", kcmp(me, sem, KCMP_SYSVSEM, 0, 0) == 0);
    CHECK("undo-list-made-for-creator", ordered(me, t, KCMP_SYSVSEM, 0, 0));
    CHECK("undo-lists-absent-equal", kcmp(t, copy, KCMP_SYSVSEM, 0, 0) == 0);
    stop_all();
}

static void epoll_items(void) {
    pid_t me = getpid();
    int ep = epoll_create1(0), fds[2];
    CHECK("pipe", pipe(fds) == 0);
    int r = fds[0], w = fds[1];
    struct epoll_event ev = {.events = EPOLLIN};
    CHECK("epoll-add", epoll_ctl(ep, EPOLL_CTL_ADD, r, &ev) == 0);
    struct { uint32_t efd, tfd, toff; } slot = {ep, r, 0};
    CHECK("epoll-item", kcmp(me, me, KCMP_EPOLL_TFD, r, (unsigned long)&slot) == 0);
    long other = kcmp(me, me, KCMP_EPOLL_TFD, w, (unsigned long)&slot);
    CHECK("epoll-item-other", other == 1 || other == 2);
    CHECK_ERR("epoll-slot-first", kcmp(me, me, KCMP_EPOLL_TFD, 999, (unsigned long)BAD), EFAULT);
    CHECK_ERR("epoll-descriptor", kcmp(me, me, KCMP_EPOLL_TFD, 999, (unsigned long)&slot), EBADF);
    slot.efd = 999;
    CHECK_ERR("epoll-instance-none", kcmp(me, me, KCMP_EPOLL_TFD, r, (unsigned long)&slot), EBADF);
    slot.efd = r;
    CHECK_ERR("epoll-not-instance", kcmp(me, me, KCMP_EPOLL_TFD, r, (unsigned long)&slot), EINVAL);
    slot.efd = ep;
    slot.toff = 1;
    CHECK_ERR("epoll-offset", kcmp(me, me, KCMP_EPOLL_TFD, r, (unsigned long)&slot), ENOENT);
    slot.toff = 0;
    slot.tfd = w;
    CHECK_ERR("epoll-not-added", kcmp(me, me, KCMP_EPOLL_TFD, r, (unsigned long)&slot), ENOENT);
    /* Two descriptions under one number: the read end stays in the set
     * through a duplicate once its number is reused. */
    int keep = dup(r);
    close(r);
    CHECK("reuse", dup(w) == r);
    CHECK("epoll-add-again", epoll_ctl(ep, EPOLL_CTL_ADD, r, &ev) == 0);
    long k[3], v[3];
    slot.tfd = r;
    for (int i = 0; i < 3; i++) {
        slot.toff = i;
        errno = 0;
        k[i] = kcmp(me, me, KCMP_EPOLL_TFD, keep, (unsigned long)&slot);
        v[i] = kcmp(me, me, KCMP_EPOLL_TFD, w, (unsigned long)&slot);
    }
    CHECK("epoll-two-items", (k[0] == 0) != (k[1] == 0) && (v[0] == 0) != (v[1] == 0) &&
                                 (k[0] == 0) != (v[0] == 0));
    CHECK("epoll-third-none", k[2] == -1 && v[2] == -1 && errno == ENOENT);
    close(keep);
    close(r);
    close(w);
    close(ep);
}

/* A leader that exited while a thread runs keeps only its signal
 * handlers. */
static int leader_word;

static void *after_leader(void *arg) {
    (void)arg;
    while (__atomic_load_n(&leader_word, __ATOMIC_ACQUIRE) != 0)
        syscall(SYS_futex, &leader_word, FUTEX_WAIT, leader_word, NULL);
    usleep(100000);
    pid_t me = getpid(), self = gettid();
    CHECK("exited-leader-vm", ordered(me, self, KCMP_VM, 0, 0));
    CHECK("exited-leader-files", ordered(me, self, KCMP_FILES, 0, 0));
    CHECK("exited-leader-fs", ordered(me, self, KCMP_FS, 0, 0));
    CHECK("exited-leader-sighand", kcmp(me, self, KCMP_SIGHAND, 0, 0) == 0);
    /* The thread has a copy of the context threads() gave a priority,
     * which fork and pthread_create passed on (copy_io). */
    CHECK("exited-leader-io", ordered(me, self, KCMP_IO, 0, 0));
    CHECK("exited-leader-undo-list", ordered(me, self, KCMP_SYSVSEM, 0, 0));
    CHECK_ERR("exited-leader-file", kcmp(me, self, KCMP_FILE, 1, 1), EBADF);
    fflush(stdout);
    exit(failures ? 1 : 0);
}

static void exited_leader(void) {
    fflush(stdout);
    pid_t c = fork();
    if (c == 0) {
        leader_word = 1;
        syscall(SYS_set_tid_address, &leader_word);
        pthread_t t;
        pthread_create(&t, NULL, after_leader, NULL);
        syscall(SYS_exit, 0);
    }
    int status = 0;
    waitpid(c, &status, 0);
    if (!WIFEXITED(status) || WEXITSTATUS(status) != 0) failures++;
}

/* Another user's process: init. */
static void others(void) {
    pid_t me = getpid();
    CHECK_ERR("init", kcmp(me, 1, KCMP_VM, 0, 0), EPERM);
    CHECK_ERR("init-before-type", kcmp(1, me, 99, 0, 0), EPERM);
}

int main(void) {
    fflush(stdout);
    pid_t c = fork();
    if (c == 0) {
        if (getuid() == 0 && (setgid(65534) != 0 || setuid(65534) != 0)) {
            printf("FAIL drop-privileges\n");
            exit(1);
        }
        order_and_files();
        threads();
        epoll_items();
        others();
        exited_leader();
        fflush(stdout);
        exit(failures ? 1 : 0);
    }
    int status = 0;
    waitpid(c, &status, 0);
    failures = !WIFEXITED(status) || WEXITSTATUS(status) != 0;
    FINISH();
}

File: src/user/linux/tests/kcmp.rs
//! `kcmp` against `kernel/kcmp.c` (Linux 6.19), through the system call on
//! every ABI: the checks in order, open file descriptions, the objects
//! threads share or keep apart (address space, tables, signal handlers,
//! I/O contexts, semaphore undo lists), epoll items, an exited leader, and
//! another process. Every expectation was checked on a Linux 7.0 kernel.

use super::harness::{Harness, P, each_abi};
use crate::user::linux::abi::Sysno;
use crate::user::linux::abi::errno_table::*;
use crate::user::linux::syscall::thread::cf::*;

const KCMP_FILE: u64 = 0;
const KCMP_VM: u64 = 1;
const KCMP_SIGHAND: u64 = 4;
const KCMP_IO: u64 = 5;
const KCMP_SYSVSEM: u64 = 6;
const KCMP_EPOLL_TFD: u64 = 7;
const THREAD: u64 = CLONE_VM | CLONE_FS | CLONE_FILES | CLONE_SIGHAND | CLONE_THREAD;
const O_RDONLY: u64 = 0;
const O_PATH: u64 = 0o10_000_000;
const AT_FDCWD: u64 = -100i64 as u64;
/// Never mapped (below `mmap_min_addr`).
const BAD: u64 = 0x10;

fn kcmp(h: &mut Harness, p1: u64, p2: u64, ty: u64, i1: u64, i2: u64) -> i64 {
    h.call(Sysno::Kcmp, &[p1, p2, ty, i1, i2])
}

/// Opens `/` with `flags`.
fn open_root(h: &mut Harness, flags: u64) -> u64 {
    let at = h.scratch;
    h.proc.state.space.write_raw(at, b"/\0").unwrap();
    h.ok(Sysno::Openat, &[AT_FDCWD, at, flags, 0])
}

/// A distinct result for distinct objects: 1 or 2, the other way round
/// when the order of the arguments is.
fn ordered(h: &mut Harness, p1: u64, p2: u64, ty: u64, i1: u64, i2: u64) -> bool {
    let (x, y) = (kcmp(h, p1, p2, ty, i1, i2), kcmp(h, p2, p1, ty, i2, i1));
    matches!((x, y), (1, 2) | (2, 1))
}

#[test]
fn kcmp_checks_in_the_kernels_order_and_compares_files() {
    // Both tasks are found (ESRCH), both may be inspected (EPERM), then
    // the type decides (EINVAL for an unknown one).
    each_abi(|abi| {
        let mut h = Harness::new(abi);
        let me = h.proc.state.pid as u64;
        assert_eq!(kcmp(&mut h, 99999, me, 99, 0, 0), -(ESRCH as i64));
        assert_eq!(kcmp(&mut h, me, 0, 99, 0, 0), -(ESRCH as i64));
        assert_eq!(kcmp(&mut h, me, me, 8, 0, 0), -(EINVAL as i64));
        assert_eq!(kcmp(&mut h, me, me, u64::MAX, 0, 0), -(EINVAL as i64));
        let a = open_root(&mut h, O_RDONLY);
        let b = open_root(&mut h, O_RDONLY);
        let d = h.ok(Sysno::Dup, &[a]);
        let p = open_root(&mut h, O_PATH);
        assert_eq!(kcmp(&mut h, me, me, KCMP_FILE, a, d), 0, "one description");
        assert!(ordered(&mut h, me, me, KCMP_FILE, a, b));
        assert!(ordered(&mut h, me, me, KCMP_FILE, a, p), "O_PATH counts");
        // The index is an unsigned int to fget_task.
        assert_eq!(kcmp(&mut h, me, me, KCMP_FILE, a, (1 << 32) | d), 0);
        assert_eq!(kcmp(&mut h, me, me, KCMP_FILE, a, 999), -(EBADF as i64));
        assert_eq!(
            kcmp(&mut h, me, me, KCMP_FILE, u64::MAX, a),
            -(EBADF as i64)
        );
        // The order is transitive over three descriptions.
        let c = open_root(&mut h, O_RDONLY);
        let mut v = [a, b, c];
        v.sort_by(|&x, &y| match kcmp(&mut h, me, me, KCMP_FILE, x, y) {
            0 => std::cmp::Ordering::Equal,
            1 => std::cmp::Ordering::Less,
            _ => std::cmp::Ordering::Greater,
        });
        assert_eq!(kcmp(&mut h, me, me, KCMP_FILE, v[0], v[2]), 1);
        // Another process may not be inspected, whichever side it is on.
        h.proc.state.config.processes = true;
        let other = u64::from(std::os::unix::process::parent_id());
        assert_eq!(kcmp(&mut h, me, other, KCMP_VM, 0, 0), -(EPERM as i64));
        assert_eq!(kcmp(&mut h, other, me, 99, 0, 0), -(EPERM as i64));
    });
}

#[test]
fn threads_share_what_their_clone_flags_shared() {
    // Threads share the address space, tables, and signal handlers; an
    // I/O context or undo list is absent (equal) until made, shared by
    // CLONE_IO or CLONE_SYSVSEM, and otherwise each task's own.
    each_abi(|abi| {
        let mut h = Harness::new(abi);
        let me = h.proc.state.pid as u64;
        let t = h.ok(Sysno::Clone, &[THREAD, 0, 0, 0, 0]);
        for ty in 1..=6 {
            assert_eq!(kcmp(&mut h, me, t, ty, 0, 0), 0, "type {ty}");
        }
        h.ok(Sysno::IoprioSet, &[1, 0, (2 << 13) | 4]);
        assert!(
            ordered(&mut h, me, t, KCMP_IO, 0, 0),
            "a context of its own"
        );
        let io = h.ok(Sysno::Clone, &[THREAD | CLONE_IO, 0, 0, 0, 0]);
        assert_eq!(kcmp(&mut h, me, io, KCMP_IO, 0, 0), 0);
        let copy = h.ok(Sysno::Clone, &[THREAD, 0, 0, 0, 0]);
        assert!(ordered(&mut h, me, copy, KCMP_IO, 0, 0), "a copy");
        let sem = h.ok(Sysno::Clone, &[THREAD | CLONE_SYSVSEM, 0, 0, 0, 0]);
        assert_eq!(kcmp(&mut h, me, sem, KCMP_SYSVSEM, 0, 0), 0, "shared");
        assert!(ordered(&mut h, me, t, KCMP_SYSVSEM, 0, 0), "made for me");
        // The leader gone, only its signal handlers are still shared.
        assert_eq!(h.start(0, Sysno::Exit, &[0]), None);
        let w = h.index_of(t as i32);
        for ty in [KCMP_VM, 2, 3, KCMP_IO, KCMP_SYSVSEM] {
            let r = h.start(w, Sysno::Kcmp, &[me, sem, ty, 0, 0]);
            assert!(matches!(r, Some(1 | 2)), "type {ty}: {r:?}");
        }
        let r = h.start(w, Sysno::Kcmp, &[me, sem, KCMP_SIGHAND, 0, 0]);
        assert_eq!(r, Some(0));
        let r = h.start(w, Sysno::Kcmp, &[me, t, KCMP_FILE, 0, 0]);
        assert_eq!(r, Some(-(EBADF as i64)), "no descriptors");
    });
}

#[test]
fn epoll_targets_are_found_by_descriptor_and_offset() {
    // kcmp_epoll_target: the slot is read first (EFAULT), then the
    // descriptors (EBADF), then the instance (EINVAL, ENOENT); items with
    // the same descriptor number come in the kernel tree's order.
    each_abi(|abi| {
        let mut h = Harness::new(abi);
        let me = h.proc.state.pid as u64;
        let m = h.anon(P, 3, false);
        let ep = h.ok(Sysno::EpollCreate1, &[0]);
        let fds = m + 0x100;
        h.ok(Sysno::Pipe2, &[fds, 0]);
        let mut b = [0u8; 8];
        h.proc.state.space.read_raw(fds, &mut b).unwrap();
        let (r, w) = (
            u64::from(u32::from_le_bytes(b[..4].try_into().unwrap())),
            u64::from(u32::from_le_bytes(b[4..].try_into().unwrap())),
        );
        let ev = m + 0x200;
        h.proc.state.space.write_raw(ev, &[1, 0, 0, 0]).unwrap();
        h.ok(Sysno::EpollCtl, &[ep, 1, r, ev]);
        let slot = m + 0x300;
        let put = |h: &Harness, efd: u64, tfd: u64, toff: u32| {
            let mut v = (efd as u32).to_le_bytes().to_vec();
            v.extend_from_slice(&(tfd as u32).to_le_bytes());
            v.extend_from_slice(&toff.to_le_bytes());
            h.proc.state.space.write_raw(slot, &v).unwrap();
        };
        put(&h, ep, r, 0);
        assert_eq!(kcmp(&mut h, me, me, KCMP_EPOLL_TFD, r, slot), 0);
        let res = kcmp(&mut h, me, me, KCMP_EPOLL_TFD, w, slot);
        assert!(matches!(res, 1 | 2));
        assert_eq!(
            kcmp(&mut h, me, me, KCMP_EPOLL_TFD, 999, BAD),
            -(EFAULT as i64)
        );
        assert_eq!(
            kcmp(&mut h, me, me, KCMP_EPOLL_TFD, 999, slot),
            -(EBADF as i64)
        );
        put(&h, 999, r, 0);
        assert_eq!(
            kcmp(&mut h, me, me, KCMP_EPOLL_TFD, r, slot),
            -(EBADF as i64)
        );
        put(&h, r, r, 0);
        assert_eq!(
            kcmp(&mut h, me, me, KCMP_EPOLL_TFD, r, slot),
            -(EINVAL as i64)
        );
        put(&h, ep, r, 1);
        assert_eq!(
            kcmp(&mut h, me, me, KCMP_EPOLL_TFD, r, slot),
            -(ENOENT as i64)
        );
        put(&h, ep, w, 0);
        assert_eq!(
            kcmp(&mut h, me, me, KCMP_EPOLL_TFD, r, slot),
            -(ENOENT as i64)
        );
        // Two descriptions under one number: the pipe's read end stays in
        // the set through a duplicate after its number is reused.
        let keep = h.ok(Sysno::Dup, &[r]);
        h.ok(Sysno::Close, &[r]);
        let again = h.ok(Sysno::Dup, &[w]);
        assert_eq!(again, r, "the lowest free number");
        h.ok(Sysno::EpollCtl, &[ep, 1, r, ev]);
        let mut found = Vec::new();
        for toff in 0..3 {
            put(&h, ep, r, toff);
            found.push((
                kcmp(&mut h, me, me, KCMP_EPOLL_TFD, keep, slot),
                kcmp(&mut h, me, me, KCMP_EPOLL_TFD, w, slot),
            ));
        }
        let e = -(ENOENT as i64);
        assert_eq!(found[2], (e, e));
        let hits: Vec<_> = found[..2].iter().map(|&(k, w)| (k == 0, w == 0)).collect();
        assert!(
            hits == [(true, false), (false, true)] || hits == [(false, true), (true, false)],
            "{found:?}"
        );
    });
}
